### ISSUE TITLE
Sanitize state persistence in sbt plugin

### ISSUE THREAD

**Issue by gzm0** [scala-js/scala-js#2171]
Historically, we have used settings to persist things (notably the incremental optimizer/linker) between multiple sbt runs (but per scope).

However, this hack did not allow us to clean up once values got replaced (n.b. on reload).

We should fix this and ensure we do it properly.

**Comment by gzm0** [comment]
I should take a look at this again, since with the advent of #2645 (improved test adapter) this will become even more relevant.

@eed3si9n You might be able to help me here: What we need is a clean way to create, use and clean up resources on a per-scope basis. We currently abuse settings and finalizers for this: We store the resource handle in a setting, so it is persistent across interactive invocations and we rely on its finalizer to clean up if a command reloads the setting (see [example][1]). This is of course sub-optimal since resources might be left dangling. Is there a better way to do this?

As a bonus point: We would like to emit statistics once an sbt execution (i.e. a full graph execution) has terminated. Is there a way to do that (same [example][1] shows our current hack to do this).

[1]: https://github.com/scala-js/scala-js/blob/9c35263b29ff84143180a1b8b5c241b60f3763d0/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala#L344-L351

**Comment by sjrd** [comment]
For the bonus point, I think that's exactly what the `onComplete` setting provides. See in https://github.com/sbt/sbt/blob/1.0.x/main/src/main/scala/sbt/Keys.scala:
```scala
val onComplete = SettingKey[() => Unit]("on-complete", "Hook to run when task evaluation completes.  The type of this setting is subject to change, pending the resolution of SI-2915.", DSetting)
```

**Comment by gzm0** [comment]
Assigning this to v1.0.0-M2 to remember that we should try extra hard to remove all internal keys. Candidates are:

- `scalaJSClearCacheStats`
- `scalaJSIRCacheHolder`

**Comment by sjrd** [comment]
Fixed in 6bbb4ee076698beccf4575518c344e60058df7d9.
